### PR TITLE
deploy: Enable `extraCreateMetadata` for provisioner

### PR DESCRIPTION
**Pull request by humblec**
Pass extra metadata when creating volumes makes it possible to detect the Kubernetes Namespace where the PVC is created. This can then be used for fetching the Vault Token (or other configuration) from the namespace where the PVC lives.

The extra metadata is available in the parameters of the CreateVolume request, for example:
```
GRPC call: /csi.v1.Controller/CreateVolume
GRPC request: {"accessibility_requirements":{"preferred":[{"segments":{"topology.rbd.csi.ceph.com/region":"testregion","topology.rbd.csi.ceph.com/zone":"testzone"}}],"requisite":[{"segments":{"topology.rbd.csi.ceph.com/region":"testregion","topology.rbd.csi.ceph.com/zone":"testzone"}}]},"capacity_range":{"required_bytes":1073741824},"name":"pvc-a18380b9-d9b7-46df-be26-693e0530c715","parameters":{"clusterID":"fa4b2e2b-ceb9-4db2-a96b-162a815a31a9","csi.storage.k8s.io/pv/name":"pvc-a18380b9-d9b7-46df-be26-693e0530c715","csi.storage.k8s.io/pvc/name":"rbd-pvc","csi.storage.k8s.io/pvc/namespace":"rbd-7887","encrypted":"true","imageFeatures":"layering","pool":"replicapool"},"secrets":"***stripped***","volume_capabilities":[{"AccessType":{"Mount":{"fs_type":"ext4","mount_flags":["discard"]}},"access_mode":{"mode":1}}]}
```

See-also: #1305

### CODE REVIEW

File: deploy/cephfs/kubernetes/csi-cephfsplugin-provisioner.yaml
Comment (nixpanic) [inline deploy/cephfs/kubernetes/csi-cephfsplugin-provisioner.yaml:44]: Missing closing `"`.

**Review by nixpanic — DISMISSED** [review]
Please correct the subject of the commit message, no trailing `:`

File: deploy/cephfs/kubernetes/csi-cephfsplugin-provisioner.yaml
Comment (humblec) [reply deploy/cephfs/kubernetes/csi-cephfsplugin-provisioner.yaml]: added.

**Review by Madhu-1 — APPROVED** [review]

**Comment by humblec** [comment]
Adding `DNM` , just to revisit the doc updates if required any.

**Comment by humblec** [comment]
> Adding `DNM` , just to revisit the doc updates if required any.

Looks like we can leave without doc update. Removing `DNM` .

**Comment by humblec** [comment]
@nixpanic ptal .

**Review by nixpanic — COMMENTED** [review]
Does there need to be an e2e tests to verify that the expected metadata is indeed available?
It would help to link to the documentation of the option in the commit message.

**Review by Madhu-1 — DISMISSED** [review]
Holding this one as we have not decided yet to go with pvc-name+namespace name for DR usecase

**Comment by humblec** [comment]
> Holding this one as we have not decided yet to go with pvc-name+namespace name for DR usecase

@Madhu-1 ideally this is not related to DR use case or naming schema. This is independent to one use case. I dont see a point in having it blocked because of that.

**Comment by Madhu-1** [comment]
> > Holding this one as we have not decided yet to go with pvc-name+namespace name for DR usecase
> 
> @Madhu-1 ideally this is not related to DR use case or naming schema. This is independent to one use case. I dont see a point in having it blocked because of that.

can you please update the same details in the PR description and also provide detail on how this can be used? I don't see a value in just  passing the metadata to CSI driver without any valid use-case

**Comment by humblec** [comment]
> > > Holding this one as we have not decided yet to go with pvc-name+namespace name for DR usecase
> > 
> > 
> > @Madhu-1 ideally this is not related to DR use case or naming schema. This is independent to one use case. I dont see a point in having it blocked because of that.
> 
> can you please update the same details in the PR description and also provide detail on how this can be used? I don't see a value in just passing the metadata to CSI driver without any valid use-case

@Madhu-1 updated the commit message and PR description with some more details.

**Comment by humblec** [comment]
@Madhu-1 @nixpanic ptal :+1:

**Review by nixpanic — DISMISSED** [review]
If this is an enablement as part of the the design discussed in #1305, we should probably wait with merging until the design has been approved and merged.

**Comment by humblec** [comment]
> If this is an enablement as part of the the design discussed in #1305, we should probably wait with merging until the design has been approved and merged.

@nixpanic this is not tied to those features in full context. It could be a useful input for CSI driver to consume as mentioned in the PR description , be it vault enhancement or rbd mirroring or more.. But this is an independent one which dont harm anything and we could consume accordingly. so IMO, we could get this in without any concerns.

**Review by nixpanic — CHANGES_REQUESTED** [review]
> @nixpanic this is not tied to those features in full context. It could be a useful input for CSI driver to consume as mentioned in the PR description , be it vault enhancement or rbd mirroring or more.. But this is an independent one which dont harm anything and we could consume accordingly. so IMO, we could get this in without any concerns.

In that case, this is just a stand-alone feature. Please add something to the documentation so that users (and developers) know it exists. It also needs e2e testing so that we can detect unintentional breakage.

**Comment by humblec** [comment]
> > @nixpanic this is not tied to those features in full context. It could be a useful input for CSI driver to consume as mentioned in the PR description , be it vault enhancement or rbd mirroring or more.. But this is an independent one which dont harm anything and we could consume accordingly. so IMO, we could get this in without any concerns.
> 
> In that case, this is just a stand-alone feature. Please add something to the documentation so that users (and developers) know it exists. It also needs e2e testing so that we can detect unintentional breakage.

@nixpanic I failed to get it , this is not a feature which is user facing! its just that, CSI driver get the additional metadata, thats it and its internal in all the sense. 
I believe the documentation or other user facing things will come in ONLY when CSI driver started to consume it for features like RBD mirroring  or Vault ..etc.. Happy to chat and explain :)

**Comment by humblec** [comment]
@Madhu-1 please take a look at this version. I have updated the PR description with the details.

**Comment by Madhu-1** [comment]
Am not going with the approach mentioned in https://github.com/ceph/ceph-csi/pull/1305 for rbd mirroring.

**Comment by nixpanic** [comment]
> @nixpanic I failed to get it , this is not a feature which is user facing! its just that, CSI driver get the additional metadata, thats it and its internal in all the sense.

Right, but there are no requirements to have this metadata at this time? And there are no consumers within Ceph-CSI that use this metadata. No e2e tests that verify the metadata is passed, is stored or retrieved anywhere. This currently looks like bloat that is not necessary.

**Comment by humblec** [comment]
> > @nixpanic I failed to get it , this is not a feature which is user facing! its just that, CSI driver get the additional metadata, thats it and its internal in all the sense.
> 
> Right, but there are no requirements to have this metadata at this time? And there are no consumers within Ceph-CSI that use this metadata.
@nixpanic it is true that, there is NO fixed or confirmed consumer at this time from our CSI driver. I agree to that.!

> No e2e tests that verify the metadata is passed, is stored or retrieved anywhere. This currently looks like bloat that is not necessary.

The E2E tests are part of the sidecar ( here external-provisioner) and its well qualified against the CSI driver. 
Unfortunately I dont understand `stored` and `retrieved` -> this is on the consuming part. Again, it depends on the feature it consume.
 
Regardless, one of the thought I have here is `recording` namespace of a PVC  in volumeContext of a PV for better filtering of the request in later cycle.

**Comment by Madhu-1** [comment]
> > > @nixpanic I failed to get it , this is not a feature which is user facing! its just that, CSI driver get the additional metadata, thats it and its internal in all the sense.
> > 
> > 
> > Right, but there are no requirements to have this metadata at this time? And there are no consumers within Ceph-CSI that use this metadata.
> 
> @nixpanic it is true that, there is NO fixed or confirmed consumer at this time from our CSI driver. I agree to that.!
> 
> > No e2e tests that verify the metadata is passed, is stored or retrieved anywhere. This currently looks like bloat that is not necessary.
> 
> The E2E tests are part of the sidecar ( here external-provisioner) and its well qualified against the CSI driver.
> Unfortunately I dont understand `stored` and `retrieved` -> this is on the consuming part. Again, it depends on the feature it consume.
> 
> Regardless, one of the thought I have here is `recording` namespace of a PVC in volumeContext of a PV for better filtering of the request in later cycle.

How feasible it is to record the namespace in PV? PV is not bound to any namespace or the PVC. a user can statically bind the PV to any pvc on any namespace

**Comment by humblec** [comment]
>How feasible it is to record the namespace in PV? PV is not bound to any namespace or the PVC. a user can statically bind the PV to any pvc on any namespace

I added some more data to https://github.com/ceph/ceph-csi/pull/1093/#issuecomment-715206813 in between. Any way to the above,  PV is a global object , but when its bound it always have a 1:1 mapping to a PVC, and PVC is namespace scoped.  Does that answer ?

**Comment by Madhu-1** [comment]
> > How feasible it is to record the namespace in PV? PV is not bound to any namespace or the PVC. a user can statically bind the PV to any pvc on any namespace
> 
> I added some more data to #1093 in between. Any way to the above, PV is a global object , but when its bound it always have a 1:1 mapping to a PVC, and PVC is namespace scoped. Does that answer ?

what if the user deletes the PVC and retain the PV and bind it to another PVC in a different namespace?

**Comment by humblec** [comment]
> > > How feasible it is to record the namespace in PV? PV is not bound to any namespace or the PVC. a user can statically bind the PV to any pvc on any namespace
> > 
> > 
> > I added some more data to #1093 in between. Any way to the above, PV is a global object , but when its bound it always have a 1:1 mapping to a PVC, and PVC is namespace scoped. Does that answer ?
> 
> what if the user deletes the PVC and retain the PV and bind it to another PVC in a different namespace?

Unfortunately I dont understand this :(.. In this case, namespace is different and NOT what we have recorded, Isnt it ?

**Comment by Madhu-1** [comment]
> > > > How feasible it is to record the namespace in PV? PV is not bound to any namespace or the PVC. a user can statically bind the PV to any pvc on any namespace
> > > 
> > > 
> > > I added some more data to #1093 in between. Any way to the above, PV is a global object , but when its bound it always have a 1:1 mapping to a PVC, and PVC is namespace scoped. Does that answer ?
> > 
> > 
> > what if the user deletes the PVC and retain the PV and bind it to another PVC in a different namespace?
> 
> Unfortunately I dont understand this :(.. In this case, namespace is different and NOT what we have recorded, Isnt it ?

Yes.

**Comment by humblec** [comment]
> > > > > How feasible it is to record the namespace in PV? PV is not bound to any namespace or the PVC. a user can statically bind the PV to any pvc on any namespace
> > > > 
> > > > 
> > > > I added some more data to #1093 in between. Any way to the above, PV is a global object , but when its bound it always have a 1:1 mapping to a PVC, and PVC is namespace scoped. Does that answer ?
> > > 
> > > 
> > > what if the user deletes the PVC and retain the PV and bind it to another PVC in a different namespace?
> > 
> > 
> > Unfortunately I dont understand this :(.. In this case, namespace is different and NOT what we have recorded, Isnt it ?
> 
> Yes.

So, its a static PV workflow and not part of  dynamic provisioning workflow. Also, this gives a `validation checkpoint` if needed later while we are operating on such a conflicting PV and if at all the requested operation depend on the `namespace` specific decisions. If its not specific or not depend on the namespace specific operation, we are free to discard it. !

**Comment by humblec** [comment]
@nixpanic @Madhu-1 are we converging to the point that, we dont have to consider this unless there is a consumer for this ? please confirm

**Comment by nixpanic** [comment]
> @nixpanic @Madhu-1 are we converging to the point that, we dont have to consider this unless there is a consumer for this ? please confirm

Yes, I would say so. We should not add features that have no clear plan yet of how they benefit the project. Once there is a requirement and way to verify the functioning of this feature, it can be added of course.

**Comment by nixpanic** [comment]
Rebased on top of the master branch and added a 1st part of a consumer. The RBD image now has an `Owner` that is set by the extra `csi.storage.k8s.io/pvc/namespace` metadata.

The [project to add support for a KMS with Vault Tokens](https://github.com/ceph/ceph-csi/projects/8) has more information on for this requirement.

**Review by nixpanic — DISMISSED** [review]

**Review by nixpanic — APPROVED** [review]

**Review by nixpanic — DISMISSED** [review]

**Comment by nixpanic** [comment]
**ci/centos/mini-e2e-helm/k8s-1.18** and **ci/centos/mini-e2e-helm/k8s-1.19** failed. It seems the Helm chart was not updated with the additional parameter for the sidecar :man_facepalming:

**Review by nixpanic — APPROVED** [review]
@Madhu-1 this is ready for review. In case the `rbdVolume.Owner` needs to be stored as RBD Image metadata (to be confirmed at a later time in #1743), I'll do that in a follow-up PR.

File: e2e/rbd_helper.go
Comment (Madhu-1) [inline e2e/rbd_helper.go:226]: can we validate the owner's name? instead of checking the key, best would be to validate the value we store

File: internal/rbd/rbd_util.go
Comment (Madhu-1) [inline internal/rbd/rbd_util.go:829]: Is this needed only when encryption is enabled? if this is not passed we are going to get Debug logs for all volume create requests.

File: e2e/rbd.go
Comment (Madhu-1) [inline e2e/rbd.go:263]: same tests need to be added for the rados namespace test also

**Review by Madhu-1 — COMMENTED** [review]
small nit, changes looks good

File: internal/rbd/rbd_util.go
Comment (nixpanic) [reply internal/rbd/rbd_util.go:829]: It is not required (at the moment) for non-encrypted volumes. The option to pass extra metadata is passed with all deployments, as we can not  know in advance if encryption will be used, or not. So the owner will be available everywhere.

In production environments, debug logging should not be enabled. I'd expect only errors (and higher) to be logged in production environments by default.

File: e2e/rbd_helper.go
Comment (nixpanic) [reply e2e/rbd_helper.go:226]: Sure, will adapt it.

File: e2e/rbd.go
Comment (nixpanic) [reply e2e/rbd.go:263]: ok, will do!

**Review by nixpanic — APPROVED** [review]

**Review by nixpanic — APPROVED** [review]

**Review by nixpanic — APPROVED** [review]

File: internal/cephfs/fsjournal.go
Comment (Madhu-1) [inline internal/cephfs/fsjournal.go:235]: the flag `extra-create-metadata` is set for both cephfs and rbd deployment above but we are not setting the owner for cephfs is there any reason for that one?

File: internal/journal/voljournal.go
Comment (Madhu-1) [inline internal/journal/voljournal.go:189]: is this needed for a snapshot also?

File: internal/rbd/rbd_util.go
Comment (Madhu-1) [reply internal/rbd/rbd_util.go:823]: in storageclass "encrypted" will be set to true right based on that we can do this operation?

File: internal/cephfs/fsjournal.go
Comment (nixpanic) [reply internal/cephfs/fsjournal.go:235]: It has been added to have the options in sync as much as possible. CephFS does not set the KMS-ID either yet, at that point the Owner can be added too.

File: internal/journal/voljournal.go
Comment (nixpanic) [reply internal/journal/voljournal.go:189]: I'd say it is needed where the `csi.volume.encryptKMS` is used, otherwise the options to would not be complete.

File: internal/rbd/rbd_util.go
Comment (nixpanic) [reply internal/rbd/rbd_util.go:823]: We could, but in the end it is just an other attribute of the PVC and RBD-image. It does not hurt to have it (and makes testing of the rados key/value simpler).

**Review by Madhu-1 — APPROVED** [review]

**Pull request merged**
